filterVisibility behaviour change in v1.8.4

<!--
If you have all reproduction steps with a complete sample app, please share as many details as possible in the sections below.

Make sure that you tried using the latest TestCafe version (https://github.com/DevExpress/testcafe/releases), where this behavior might have been already addressed.

Before submitting an issue, please check CONTRIBUTING.md and existing issues in this repository (https://github.com/DevExpress/testcafe/issues) in case a similar issue exists or was already addressed. This may save your time (and ours).
-->

### What is your Test Scenario?
<!-- Describe what you'd like to test. -->
I upgraded to the latest version, to take advantage of #4904. However, since doing so, unrelated functionality has now broken in our regression tests.

### What is the Current behavior?
<!-- Describe the behavior you see and consider invalid. -->
 In order to reduce code duplication, we have 2 selector values for our `login` button (one is the element found when navigating to the login page, and the other is for our login modal), then use `.filterVisible()` to filter the visible element.

After upgrading, `await t.expect(loginButton.visible).ok();` now fails for some of our tests (_navigating to the login page_). Downgrading to 1.8.3 allows those tests to pass.

### What is the Expected behavior?
<!-- Describe what you expected to happen. -->
I would expect these same tests to pass in the latest version, and not have breaking changes with upgrading to the latest version. 

### What is your web application and your TestCafe test code?
<!-- Share a public accessible link to your application or provide a simple app which we can run. -->
 
<details>
<summary>Your website URL (or attach your complete example):</summary>

<!-- Provide your website URL or attach a sample. Note: if your website requires any additional access procedures like authentication, please ask the website owner to send us a written confirmation at [support@devexpress.com](mailto:support@devexpress.com) in a free text form. It will allow the DevExpress staff to remotely access the website and its internal resources for research, testing, and debugging purposes. -->
http://www.change.org
</details>

<details>
<summary>Your complete test code (or attach your test files):</summary>

<!-- Paste your test code here: --> 
```js
import { Selector } from 'testcafe';

fixture('test login functionality').page('www.change.org');

test.only('Check for login button', async t => {
  const loginButton = Selector('#content [type="submit"], .js-login-button').filterVisible();

  await t
    .navigateTo('/login_or_join')
    .expect(loginButton.visible)
    .ok();
});
```
</details>

<details>
<summary>Your complete configuration file (if any):</summary>

<!-- Paste your complete test config file here (even if it is huge): -->
```

```
</details>

<details>
<summary>Your complete test report:</summary>

<!-- Paste your complete result test report here (even if it is huge): -->
```
test login functionality
 ✖ Check for login button (screenshots: /Users/**/artifacts/screenshots/test-1/_errors/1.png)

   1) AssertionError: expected false to be truthy

      Browser: Chrome 81.0.4044.113 / macOS 10.15.4
      Screenshot: /Users/**/artifacts/screenshots/test-1/_errors/1.png

          6 |  const loginButton = Selector('#content [type="submit"], .js-login-button').filterVisible();
          7 |
          8 |  await t
          9 |    .navigateTo('/login_or_join')
         10 |    .expect(loginButton.visible)
       > 11 |    .ok();
         12 |});
         13 |

         at <anonymous> (/Users/**/tests/demo/test.js:11:6)
         at <anonymous> (/Users/**/tests/demo/test.js:5:1)
         at <anonymous> (/Users/**/node_modules/testcafe/src/api/wrap-test-function.js:17:28)
         at TestRun._executeTestFn (/Users/**/node_modules/testcafe/src/test-run/index.js:295:19)
         at TestRun.start (/Users/**/node_modules/testcafe/src/test-run/index.js:345:24)



 1/1 failed (18s)
```
</details>

<details>
<summary>Screenshots:</summary>
<!-- If applicable, add screenshots to help explain the issue. -->

```
![1](https://user-images.githubusercontent.com/37812673/79806807-24548d80-831e-11ea-9d59-a406b7e2b39e.png)

```
</details>

### Steps to Reproduce:
<!-- Describe what we should do to reproduce the behaviour you encountered. -->

1. Install v1.8.3, and run the above test
2. Upgrade to v1.8.4 and re-run test
 
### Your Environment details:
 
* testcafe version: 1.8.4                  <!-- run `testcafe -v` -->
* node.js version: 12.14.1                   <!-- run `node -v` -->
* command-line arguments: testcafe chrome:headless <!-- example: "testcafe ie,chrome -e test.js" -->
* browser name and version: Chrome 81.0.4044.113 <!-- example: IE 11, Chrome 69, Firefox 100, etc. -->
* platform and version:  MacOS 10.15.4      <!-- example: "macOS 10.14, Windows, Linux Ubuntu 18.04.1, iOS 12 -->
* other:                                   <!-- any notes you consider important -->

--- a/src/client-functions/selectors/add-api.js
+++ b/src/client-functions/selectors/add-api.js
@@ -68,9 +68,9 @@ const expandSelectorResults = (new ClientFunctionBuilder((selector, populateDeri
 
 })).getFunction();
 
-async function getSnapshot (getSelector, callsite, SelectorBuilder) {
+async function getSnapshot (getSelector, callsite, SelectorBuilder, getVisibleValueMode) {
     let node       = null;
-    const selector = new SelectorBuilder(getSelector(), { needError: true }, { instantiation: 'Selector' }).getFunction();
+    const selector = new SelectorBuilder(getSelector(), { getVisibleValueMode, needError: true }, { instantiation: 'Selector' }).getFunction();
 
     try {
         node = await selector();
@@ -126,11 +126,12 @@ function addSnapshotProperties (obj, getSelector, SelectorBuilder, properties) {
 function addVisibleProperty ({ obj, getSelector, SelectorBuilder }) {
     Object.defineProperty(obj, VISIBLE_PROP_NAME, {
         get: () => {
-            return ReExecutablePromise.fromFn(async () => {
-                const builder  = new SelectorBuilder(getSelector(), { getVisibleValueMode: true }, { instantiation: 'Selector' });
-                const resultFn = builder.getFunction();
+            const callsite = getCallsiteForMethod('get');
 
-                return resultFn();
+            return ReExecutablePromise.fromFn(async () => {
+                const snapshot = await getSnapshot(getSelector, callsite, SelectorBuilder, true);
+
+                return !!snapshot && snapshot[VISIBLE_PROP_NAME];
             });
         }
     });

--- a/src/client/driver/command-executors/client-functions/selector-executor/filter.js
+++ b/src/client/driver/command-executors/client-functions/selector-executor/filter.js
@@ -54,11 +54,6 @@ class SelectorFilter {
             else
                 filtered = filtered.length;
         }
-        else if (options.getVisibleValueMode) {
-            const el = node[0];
-
-            return !!el && visible(el);
-        }
         else {
             if (options.collectionMode) {
                 if (options.index !== null) {

--- a/src/client/driver/command-executors/client-functions/selector-executor/index.js
+++ b/src/client/driver/command-executors/client-functions/selector-executor/index.js
@@ -4,7 +4,7 @@ import ClientFunctionExecutor from '../client-function-executor';
 import { exists, visible } from '../../../utils/element-utils';
 import { createReplicator, FunctionTransform, SelectorNodeTransform } from '../replicator';
 import './filter';
-import returnSinglePropMode from '../../../../../client-functions/return-single-prop-mode';
+
 
 const CHECK_ELEMENT_DELAY = 200;
 
@@ -15,7 +15,8 @@ export default class SelectorExecutor extends ClientFunctionExecutor {
         this.createNotFoundError    = createNotFoundError;
         this.createIsInvisibleError = createIsInvisibleError;
         this.timeout                = typeof command.timeout === 'number' ? command.timeout : globalTimeout;
-        this.returnSinglePropMode   = returnSinglePropMode(this.dependencies.filterOptions);
+        this.counterMode            = this.dependencies.filterOptions.counterMode;
+        this.getVisibleValueMode    = this.dependencies.filterOptions.getVisibleValueMode;
 
         if (startTime) {
             const elapsed = new Date() - startTime;
@@ -44,13 +45,16 @@ export default class SelectorExecutor extends ClientFunctionExecutor {
         return null;
     }
 
+    _getTimeoutError (elementExists) {
+        return elementExists ? this.createIsInvisibleError : this.createNotFoundError;
+    }
+
     _validateElement (args, startTime) {
         return Promise.resolve()
             .then(() => this.fn.apply(window, args))
             .then(el => {
                 const isElementExists    = exists(el);
                 const isElementVisible   = !this.command.visibilityCheck || visible(el);
-                const createTimeoutError = !isElementExists ? this.createNotFoundError : this.createIsInvisibleError;
                 const isTimeout          = new Date() - startTime >= this.timeout;
 
                 if (isElementExists && isElementVisible)
@@ -58,6 +62,8 @@ export default class SelectorExecutor extends ClientFunctionExecutor {
 
                 if (!isTimeout)
                     return delay(CHECK_ELEMENT_DELAY).then(() => this._validateElement(args, startTime));
+
+                const createTimeoutError = this.getVisibleValueMode ? null : this._getTimeoutError(isElementExists);
 
                 if (createTimeoutError)
                     throw createTimeoutError(this._getTimeoutErrorParams());
@@ -67,12 +73,12 @@ export default class SelectorExecutor extends ClientFunctionExecutor {
     }
 
     _executeFn (args) {
-        if (this.returnSinglePropMode)
+        if (this.counterMode)
             return super._executeFn(args);
 
         const startTime = new Date();
-        let error     = null;
-        let element   = null;
+        let error       = null;
+        let element     = null;
 
         return this
             ._validateElement(args, startTime)

--- a/test/functional/fixtures/api/es-next/selector/pages/visible.html
+++ b/test/functional/fixtures/api/es-next/selector/pages/visible.html
@@ -5,7 +5,7 @@
     <title>Title</title>
 </head>
 <body>
-    <div></div>
-    <div></div>
+    <div id="invisible"></div>
+    <div id="visible">text</div>
 </body>
 </html>

--- a/test/functional/fixtures/api/es-next/selector/testcafe-fixtures/selector-visible-test.js
+++ b/test/functional/fixtures/api/es-next/selector/testcafe-fixtures/selector-visible-test.js
@@ -1,7 +1,7 @@
 import { Selector } from 'testcafe';
 
 fixture `Selector.visible`
-    .page('http://localhost:3000/fixtures/api/es-next/selector/pages/visible.html');
+    .page('../pages/visible.html');
 
 test('test', async t => {
     const nonExistingElement = Selector('#wrong-selector');
@@ -10,10 +10,16 @@ test('test', async t => {
         .expect(nonExistingElement.exists).notOk()
         .expect(nonExistingElement.visible).notOk();
 
-    const existingElement = Selector('div');
+    const invisibleElement = Selector('#invisible');
 
     await t
-        .expect(existingElement.exists).ok()
-        .expect(existingElement.visible).notOk();
+        .expect(invisibleElement.exists).ok()
+        .expect(invisibleElement.visible).notOk();
+
+    const visibleElement = Selector('#invisible, #visible').filterVisible();
+
+    await t.expect(visibleElement.exists).ok();
+    await t.expect((await visibleElement()).visible).ok();
+    await t.expect(visibleElement.visible).ok();
 });
 

